proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,134 @@
+
+# Basic example
+
+The FileSimilarityRetriever would be instantiated as follows:
+
+``` python
+
+	  retriever = FileSimilarityRetriever(
+	      document_store = ElasticSearchDocumentStore,
+	      primary_retriever = EmbeddingRetriever, # defined separately, see full pipeline example below
+          top_k=30,
+	      file_aggregation_key = "file_id",
+          max_num_queries = 50
+	  )
+```
+
+And here is an example of how the node would work in the context of a full pipeline:
+
+``` yaml
+
+version: '1.19.0'
+name: 'FileSim'
+
+components:
+  - name: DocumentStore
+    type: DeepsetCloudDocumentStore
+    params:
+      similarity: cosine
+      embedding_dim: 768
+      return_embedding: false
+
+  - name: EmbeddingRetriever
+    type: EmbeddingRetriever
+    params:
+      document_store: DocumentStore
+      embedding_model: sentence-transformers/all-mpnet-base-v2
+      max_seq_len: 400
+      batch_size: 32
+      model_format: sentence_transformers
+      top_k: 50
+  - name: TextConverter # Converts files into documents
+    type: TextConverter
+  - name: FileSimilarityRetriever
+    type: FileSimilarityRetriever
+    params:
+      document_store: DocumentStore
+      primary_retriever: EmbeddingRetriever
+      top_k: 30
+      file_aggregation_key: file_id
+      max_num_queries: 50 # changed 2022-11-14
+
+pipelines:
+  - name: query
+    nodes:
+      - name: FileSimilarityRetriever
+        inputs: [Query]
+
+  - name: indexing
+    nodes:
+      - name: TextConverter
+        inputs: [File]
+
+      - name: EmbeddingRetriever
+        inputs: [TextConverter]
+
+      - name: DocumentStore
+        inputs: [EmbeddingRetriever]
+
+```
+
+# Motivation
+
+Initially this retriever was implemented by Mathis for a customer case to quickly retrieve documents similar to a given file. The reason we would like to add it to Haystack is because we see wider use for this node in future customer and community cases.
+
+One example where this could be useful is in academic writing, if you have a large number of scientific sources (journals, books, conference proceedings) stored and you'd like to find all that are similar to a specific article or to what you've written so far (in the latter case, it can also be helpful for plagiarism detection). And as for industry use cases, report writing in any field can be facilitated by file similarity retrieval. It can also be very helpful for lawyers preparing a case as well as journalists doing research for their articles.
+
+# Detailed design
+
+You can see the code for this proposal here: https://github.com/deepset-ai/haystack/pull/5666/files.
+
+As a general description, the FileSimilarityRetriever works by:
+1. Getting all documents corresponding to the provided file ID from the document store. The documents can be obtained together with their embeddings, to save resources by avoiding re-calculating them at query time.
```

Comment:
Great point! This means that no machine learning models are needed to run at query time to get these recommendations right?

Reply:
Yes, I think so